src/DataDictionary/graph/GraphCalculator/graphCalculatorHelper.js

```diff
@@ -0,0 +1,520 @@
+import Viz from 'viz.js';
+import { Module, render } from 'viz.js/full.render';
+import _ from 'underscore';
+import { createNodesAndEdges, createDotStrinByNodesEdges } from '../../../GraphUtils/utils';
+import { truncateLines, graphStyleConfig } from '../../utils';
+import { getCategoryColor } from '../../NodeCategories/helper';
+
+export const getAllTypes = (nodes) => {
+  const types = nodes.map(node => node.type);
+  const uniqueTypes = _.uniq(types);
+  return uniqueTypes;
+};
+
+export const getAllChildrenNodeIDs = (highlightingNode, nodes) => {
+  const relatedNodeIDs = [];
+  let currentLevelNodeIDs = highlightingNode.outLinks;
+  while (currentLevelNodeIDs.length > 0) {
+    const nextLevelNodeIDs = [];
+    currentLevelNodeIDs.forEach((nodeId) => {
+      if (relatedNodeIDs.includes(nodeId) || nextLevelNodeIDs.includes(nodeId)) return;
+      relatedNodeIDs.push(nodeId);
+      const originNode = nodes.find(n => (n.id === nodeId));
+      const nextLevel = originNode.outLinks;
+      nextLevel.forEach((outNodeId) => {
+        nextLevelNodeIDs.push(outNodeId);
+      });
+    });
+    currentLevelNodeIDs = nextLevelNodeIDs;
+  }
+  return relatedNodeIDs;
+};
+
+export const getAllChildrenLinks = (startingNode, nodes) => {
+  let currentLevelNodeIDs = startingNode.outLinks;
+  const relatedLinks = currentLevelNodeIDs.map(outID => ({
+    source: startingNode.id,
+    target: outID,
+  }));
+  while (currentLevelNodeIDs.length > 0) {
+    const nextLevelNodeIDs = [];
+    for (let i = 0; i < currentLevelNodeIDs.length; i += 1) {
+      const nodeID = currentLevelNodeIDs[i];
+      if (nextLevelNodeIDs.includes(nodeID)) break;
+      const originNode = nodes.find(n => (n.id === nodeID));
+      const nextLevel = originNode.outLinks;
+      for (let j = 0; j < nextLevel.length; j += 1) {
+        const outNodeID = nextLevel[j];
+        relatedLinks.push({
+          source: nodeID,
+          target: outNodeID,
+        });
+        if (nextLevelNodeIDs.includes(outNodeID)
+          || currentLevelNodeIDs.includes(outNodeID)) break;
+        nextLevelNodeIDs.push(outNodeID);
+      }
+    }
+    currentLevelNodeIDs = nextLevelNodeIDs;
+  }
+  return relatedLinks;
+};
+
+/* eslint-disable no-underscore-dangle */
+export const calculateGraphLayout = (dictionary, countsSearch, linksSearch) => {
+  const { nodes, edges } = createNodesAndEdges({
+    dictionary,
+    counts_search: countsSearch,
+    links_search: linksSearch,
+  }, true, []);
+  const dotString = createDotStrinByNodesEdges(nodes, edges);
+  const viz = new Viz({ Module, render });
+  return viz.renderJSONObject(dotString)
+    .then((renderedJSON) => {
+      // draw nodes
+      const renderedNodes = renderedJSON.objects
```

Comment:
should we separate to sub-functions for this `renderNodes` and `renderEdges` as well?